pkg/registry/parse.go

```diff
@@ -0,0 +1,253 @@
+package registry
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+)
+
+type bundleParser struct {
+	log *logrus.Entry
+}
+
+func newBundleParser(log *logrus.Entry) *bundleParser {
+	return &bundleParser{
+		log: log,
+	}
+}
+
+// Parse parses the given FS into a Bundle.
+func (b *bundleParser) Parse(root fs.FS) (*Bundle, error) {
+	if root == nil {
+		return nil, fmt.Errorf("filesystem is nil")
+	}
+
+	bundle := &Bundle{}
+	manifests, err := fs.Sub(root, "manifests")
+	if err != nil {
+		return nil, fmt.Errorf("error opening manifests directory: %s", err)
+	}
+	if err := b.addManifests(manifests, bundle); err != nil {
+		return nil, err
+	}
+
+	metadata, err := fs.Sub(root, "metadata")
+	if err != nil {
+		return nil, fmt.Errorf("error opening metadata directory: %s", err)
+	}
+	if err := b.addMetadata(metadata, bundle); err != nil {
+		return nil, err
+	}
+
+	derived, err := b.derivedProperties(bundle)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive properties: %s", err)
+	}
+
+	bundle.Properties = propertySet(append(bundle.Properties, derived...))
+
+	return bundle, nil
+}
+
+// addManifests adds the result of parsing the manifests directory to a bundle.
+func (b *bundleParser) addManifests(manifests fs.FS, bundle *Bundle) error {
+	files, err := fs.ReadDir(manifests, ".")
+	if err != nil {
+		return err
+	}
+
+	var csvFound bool
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+
+		name := f.Name()
+		if strings.HasPrefix(name, ".") {
+			continue
+		}
+
+		obj := &unstructured.Unstructured{}
+		if err = decodeFileFS(manifests, name, obj); err != nil {
+			b.log.Warnf("failed to decode: %s", err)
+			continue
+		}
+
+		// Only include the first CSV we find in the
+		if obj.GetKind() == operatorsv1alpha1.ClusterServiceVersionKind {
+			if csvFound {
+				continue
+			}
+			csvFound = true
+		}
+
+		if obj.Object != nil {
+			bundle.Add(obj)
+		}
+	}
+
+	if bundle.Size() == 0 {
+		return fmt.Errorf("no bundle objects found")
+	}
+
+	csv, err := bundle.ClusterServiceVersion()
+	if err != nil {
+		return err
+	}
+	if csv == nil {
+		return fmt.Errorf("no csv in bundle")
+	}
+
+	bundle.Name = csv.GetName()
+	if err := bundle.AllProvidedAPIsInBundle(); err != nil {
+		return fmt.Errorf("error checking provided apis in bundle %s: %s", bundle.Name, err)
+	}
+
+	return err
```

Comment:
nit: should we return `nil` at the end here?

Reply:
that's probably more idiomatic than what's currently here. I'll change that.